res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1135 @@
+        }
+
+        // common functions
+        // In most cases, you should not overwrite these.
+        inGetParameter() {
+            return engine.getParameter(this.group, this.inKey);
+        }
+        inSetParameter(value) {
+            engine.setParameter(this.group, this.inKey, value);
+        }
+        inGetValue() {
+            return engine.getValue(this.group, this.inKey);
+        }
+        inSetValue(value) {
+            engine.setValue(this.group, this.inKey, value);
+        }
+        inToggle() {
+            this.inSetValue(!this.inGetValue());
+        }
+
+        outGetParameter() {
+            return engine.getParameter(this.group, this.outKey);
+        }
+        outSetParameter(value) {
+            engine.setParameter(this.group, this.outKey, value);
+        }
+        outGetValue() {
+            return engine.getValue(this.group, this.outKey);
+        }
+        outSetValue(value) {
+            engine.setValue(this.group, this.outKey, value);
+        }
+        outToggle() {
+            this.outSetValue(!this.outGetValue());
+        }
+
+        connect() {
+            /**
+            Override this method with a custom one to connect multiple Mixxx COs for a single Component.
+            Add the connection objects to the this.connections array so they all get disconnected just
+            by calling this.disconnect(). This can be helpful for multicolor LEDs that show a
+            different color depending on the state of different Mixxx COs. Refer to
+            SamplerButton.connect() and SamplerButton.output() for an example.
+            **/
+            if (this.group !== undefined &&
+                this.outKey !== undefined &&
+                this.output !== undefined &&
+                typeof this.output === "function") {
+                this.connections[0] = engine.makeConnection(this.group, this.outKey, this.output.bind(this));
+            }
+        }
+        disconnect() {
+            for (const conn of this.connections) {
+                conn.disconnect();
+            }
+        }
+        trigger() {
+
+            for (const conn of this.connections) {
+                conn.trigger();
+            }
+        }
+        send(value) {
+            if (this.midi === undefined || this.midi[0] === undefined || this.midi[1] === undefined) {
+                return;
+            }
+            midi.sendShortMsg(this.midi[0], this.midi[1], value);
+            if (this.sendShifted) {
+                if (this.shiftChannel) {
+                    midi.sendShortMsg(this.midi[0] + this.shiftOffset, this.midi[1], value);
+                } else if (this.shiftControl) {
+                    midi.sendShortMsg(this.midi[0], this.midi[1] + this.shiftOffset, value);
+                }
+            }
+        }
+    };
+    Component.prototype.max = 127;
+    Component.prototype.shiftOffset = 0;
+    Component.prototype.sendShifted = false;
```

Comment:
I think this is where we should define this.midi so it's more clear how to use it, but perhaps this whole block should move to the top of the class.    Again, my thinking is that we want to have the parts the programmer is most likely to change at the top of the file, and the internal guts at the bottom.

Reply:
The issue is that there aren't really sensible default values for `this.midi`. Components might be used without outputting anything just to track some state or just for input handling (in the case for pots). 
We could make `this.midi` explicit by `Component.prototype.midi = [undefined, undefined]` (essentially just declaring the property)?

Reorderings in general should be fine, sure.

Reply:
`this.midi` can't be defined on the prototype. It is unique to each instance of Component.

Reply:
It can be defined there, but each instance needs to overwrite it regardless, so defining it there does not make a whole lot of sense as already pointed out.

Reply:
I once tried to move a property onto a prototype and have each instance override it, but that did not work because each instance was actually sharing the same data.

Reply:
hm ok.  This may just be a limitation of javascript then?

Reply:
> I once tried to move a property onto a prototype and have each instance override it, but that did not work because each instance was actually sharing the same data.

Thanks for the insight. I'll try to recreate this issue.

Reply:
I wouldn't say that's a limitation. That's just how prototypes work. If you don't want that, just set a normal property, not the prototype's property, which is what the code already does.